app.js

```diff
@@ -109,85 +109,79 @@ var combineQueries = (q1, q2) => {
+    console.log(cache[slug]);
+    return idOnly ? cache[slug].id : cache[slug].dest;
+  } else {
+    console.log("Oops. Can't find useful data in cache. Asking Postgres.");
+    const res = await client.query(
+      "SELECT id, destination FROM links WHERE slug = $1",
+      [slug]
+    );
+
+    if (res.rows.length > 0) {
+      cache[slug] = {
+        id: res.rows[0].id,
+        dest: res.rows[0].destination,
+        expires: timeNow + parseInt(process.env.CACHE_EXPIRATION),
+      };
+
+      return idOnly ? res.rows[0].id : res.rows[0].destination;
     } else {
-      console.log("Oops. Can't find useful data in cache. Asking Airtable.");
-      base("Links")
-        .select({
-          filterByFormula: '{slug} = "' + slug + '"',
-        })
-        .eachPage(
-          function page(records, fetchNextPage) {
-            if (records.length > 0) {
-              records.forEach(function (record) {
-                cache[slug] = {
-                  id: record.getId(),
-                  dest: record.get("destination"),
-                  expires: timeNow + parseInt(process.env.CACHE_EXPIRATION),
-                };
-
-                if (idOnly) resolve(record.getId());
-                else resolve(record.get("destination"));
-              });
-            } else {
-              fetchNextPage();
-            }
-          },
-          function done(err) {
-            if (err) {
-              // api jam
-              console.error(err);
-              reject(500);
-            } else {
-              // all records scanned - no match
-              reject(404);
-            }
-          }
-        );
+      return null;
     }
-  });
+  }
 };
 
-function logAccess(ip, ua, slug, url) {
+async function logAccess(ip, ua, slug, url) {
   if (process.env.LOGGING == "off") return;
 
-  // UA strings to identify as bot
   const botUA = ["apex/ping/v1.0"];
 
-  // do not log if the BOT_LOGGING flag is off
   if (process.env.BOT_LOGGING == "off" && (isBot(ua) || botUA.includes(ua)))
     return;
 
-  var data = {
-    Timestamp: Date.now(),
-    "Client IP": ip,
-    "User Agent": ua,
-    Bot: isBot(ua) || botUA.includes(ua),
-    Slug: [],
-    URL: url,
+  let linkId;
+  try {
+    linkId = await lookup(slug, true);
+    if (linkId === null) {
+      console.log("Slug not found, skipping logging");
+      return;
+    }
+  } catch (e) {
+    console.log(e);
+  }
+
+  const data = {
+    timestamp: new Date(),
+    client_ip: ip,
+    user_agent: ua,
+    bot: isBot(ua) || botUA.includes(ua),
+    slug: linkId,
+    url: url,
   };
 
-  lookup(slug, true)
-    .then((result) => {
-      data["Slug"][0] = result;
-    })
-    .catch(e => { console.log(e) })
-    .finally(() => {
-      base("Log").create(data, function (err, record) {
-        if (err) {
-          console.error(err);
-          return;
-        }
-      });
-    });
+  client.query(
+    `INSERT INTO logs (timestamp, client_ip, user_agent, bot, slug, url)
```

Comment:
What's the schema design here? You can likely achieve better performance by using `slug` as the primary key.

Reply:
The `slug` colum isnt primary key in the current desgin because we would need tot log multiple events with the same slug. i am aware that using `slug` as an index could speed up the queires. however from my experince i know that this could potentially slow down data insertion and updates , so lmk which fit best , i am happy to update it but for now i am just ignoring this one and waiting for the reply / feed back :)